### PR TITLE
Fix delegation and undelegation strategy with the updated val set.

### PR DESCRIPTION
**Pull request by arhamchordia**
## 1. Overview

Added delegation and undelegation strategy fix.

## 2. Implementation details

Done by generating a combined list of delegation state and allow listed validator set.

## 4. Checklist

- [ ] Does the Readme need to be updated?

### CODE REVIEW

**Review by puneet2019 — APPROVED** [review]
looks ok, but can we not reduce this function?
Looks very big and complex. Will have @Anmol1696 take a look

File: x/lscosmos/keeper/allowlisted_validators.go [hunk 1]
```diff
@@ -20,3 +20,69 @@ func (k Keeper) GetAllowListedValidators(ctx sdk.Context) types.AllowListedValid
 
 	return allowListedValidators
 }
+
+// GetAllValidatorsState returns the combined allowed listed validators set and combined
+// delegation state. It is done to keep the old validators in the loop while calculating weighted amounts
+// for delegation and undelegation
+func (k Keeper) GetAllValidatorsState(ctx sdk.Context) (types.AllowListedValidators, types.DelegationState) {
```
Comment (Anmol1696) [inline x/lscosmos/keeper/allowlisted_validators.go:27]: where is the test for this function?

Comment (arhamchordia) [reply x/lscosmos/keeper/allowlisted_validators.go:27]: Tests will be added later. It was needed urgently.

Comment (Anmol1696) [reply x/lscosmos/keeper/allowlisted_validators.go:27]: hard to figure out what the code is doing without tests

Comment (arhamchordia) [reply x/lscosmos/keeper/allowlisted_validators.go:27]: Added test file.

File: x/lscosmos/keeper/abci.go [hunk 2]
```diff
@@ -33,8 +33,7 @@ func (k Keeper) DoDelegate(ctx sdk.Context) error {
 	hostAccounts := k.GetHostAccounts(ctx)
 
 	delegatableAmount := delegationState.HostDelegationAccountBalance.AmountOf(hostChainParams.BaseDenom)
-	allowlistedValidators := k.GetAllowListedValidators(ctx)
-	if !delegatableAmount.GT(sdk.NewInt(int64(len(allowlistedValidators.AllowListedValidators)))) {
+	if !delegatableAmount.IsPositive() {
```
Comment (Anmol1696) [inline x/lscosmos/keeper/abci.go:36]: why this? i thought we needed the min value otherwise will break weight time delegation

**Review by Anmol1696 — APPROVED** [review]
LGTM for now

File: x/lscosmos/keeper/abci.go (hunk 2 above)
Comment (Anmol1696) [reply x/lscosmos/keeper/abci.go]: pushed change here

**Pull request merged**